File: scripts/strategy_logic.py
"""Shared strategy parameter and filtering logic for dashboard exports."""

from __future__ import annotations

import json
import re
from dataclasses import dataclass
from pathlib import Path
from typing import Any, Dict, Iterable, Tuple

import pandas as pd

DEFAULT_PARAMS: Dict[str, float] = {
    "home_win_rate_threshold": 0.0,
    "odds_min": 1.0,
    "odds_max": 3.2,
    "prob_threshold": 0.5,
    "min_ev": 0.0,
}
SUPPORTED_VERSION = 1


@dataclass
class StrategyParams:
    """Versioned strategy params payload."""

    version: int
    params: Dict[str, float]
    source: str


def _normalize_key(key: str) -> str:
    key = key.strip().lower()
    key = re.sub(r"[\s\-]+", "_", key)
    key = re.sub(r"[^a-z0-9_]", "", key)
    return re.sub(r"_+", "_", key)


def _coerce_float(value: Any) -> float | None:
    if value is None:
        return None
    try:
        return float(value)
    except (TypeError, ValueError):
        return None


def load_strategy_params(path: Path | None) -> StrategyParams:
    """Load versioned strategy parameters with default fallback."""
    if path is None or not path.exists():
        return StrategyParams(version=SUPPORTED_VERSION, params=DEFAULT_PARAMS.copy(), source="defaults")

    payload = json.loads(path.read_text(encoding="utf-8"))
    if not isinstance(payload, dict):
        raise ValueError(f"Invalid strategy params format in {path}")

    version = int(payload.get("version", SUPPORTED_VERSION))
    if version != SUPPORTED_VERSION:
        raise ValueError(f"Unsupported strategy params version {version} in {path}; expected {SUPPORTED_VERSION}")

    raw_params = payload.get("params", payload.get("params_used", {}))
    if not isinstance(raw_params, dict):
        raw_params = {}
    # Backward compatibility: some producers store params flat at the top-level.
    if not raw_params:
        raw_params = {
            k: v
            for k, v in payload.items()
            if _normalize_key(str(k)) in {"home_win_rate_threshold", "odds_min", "odds_max", "prob_threshold", "min_ev", "min_ev_per_100"}
        }

    normalized: Dict[str, float] = DEFAULT_PARAMS.copy()
    for key, value in raw_params.items():
        coerced = _coerce_float(value)
        if coerced is not None:
            normalized_key = _normalize_key(str(key))
            if normalized_key == "min_ev_per_100":
                normalized_key = "min_ev"
            normalized[normalized_key] = coerced

    return StrategyParams(version=version, params=normalized, source=str(path))


def apply_strategy_filters(rows: pd.DataFrame, params: StrategyParams) -> Tuple[pd.DataFrame, Dict[str, float]]:
    """Annotate strategy columns and return filtered rows plus params map."""
    frame = rows.copy()
    if frame.empty:
        return frame, params.params

    frame["prob_used"] = pd.to_numeric(frame.get("prob_used", frame.get("prob_iso")), errors="coerce")
    frame["odds_1"] = pd.to_numeric(frame.get("odds_1", frame.get("closing_home_odds")), errors="coerce")
    frame["ev_eur_per_100"] = pd.to_numeric(
        frame.get("ev_eur_per_100", frame.get("EV_€_per_100", frame.get("ev_per_100"))), errors="coerce"
    )
    frame["home_win_rate"] = pd.to_numeric(frame.get("home_win_rate"), errors="coerce")

    implied = 1.0 / frame["odds_1"]
    frame["model_market_gap"] = frame["prob_used"] - implied
    frame["model_market_gap_flag"] = frame["model_market_gap"] > 0

    blocked: list[str] = []
    for _, row in frame.iterrows():
        reasons: list[str] = []
        if pd.isna(row["prob_used"]) or row["prob_used"] < params.params["prob_threshold"]:
            reasons.append("prob_threshold")
        if pd.isna(row["odds_1"]) or row["odds_1"] < params.params["odds_min"] or row["odds_1"] > params.params["odds_max"]:
            reasons.append("odds_range")
        if pd.isna(row["ev_eur_per_100"]) or row["ev_eur_per_100"] <= params.params["min_ev"]:
            reasons.append("min_ev")
        if pd.isna(row["home_win_rate"]) or row["home_win_rate"] < params.params["home_win_rate_threshold"]:
            reasons.append("home_win_rate_threshold")
        blocked.append("|".join(reasons))

    frame["blocked_by"] = blocked
    filtered = frame[frame["blocked_by"] == ""].copy()
    return filtered, params.params


def required_columns() -> Iterable[str]:
    return (
        "date",
        "home_team",
        "away_team",
        "home_win_rate",
        "prob_iso",
        "prob_used",
        "win",
        "pnl",
    )

File: scripts/strategy_logic.mjs
import fs from "fs";

const normalizeKey = (key) =>
  key
    .trim()
    .toLowerCase()
    .replace(/[\s-]+/g, "_")
    .replace(/[^a-z0-9_]/g, "")
    .replace(/_+/g, "_");

const coerceNumber = (value) => {
  if (value === null || value === undefined) return null;
  const n = Number(value);
  return Number.isFinite(n) ? n : null;
};

const parseDate = (value) => {
  if (!value) return null;
  const trimmed = String(value).trim();
  if (!trimmed) return null;
  const date = new Date(trimmed);
  if (Number.isNaN(date.getTime())) return null;
  return date;
};

export const DEFAULT_PARAMS = {
  home_win_rate_threshold: 0,
  odds_min: 1,
  odds_max: 3.2,
  prob_threshold: 0.5,
  min_ev: 0,
};

const extractParamDict = (payload) => {
  const candidates = [];
  ["params", "params_used", "active_params", "thresholds", "strategy", "filters"].forEach((key) => {
    const value = payload?.[key];
    if (value && typeof value === "object" && !Array.isArray(value)) {
      candidates.push(value);
    }
  });
  const meta = payload?.meta;
  if (meta && typeof meta === "object" && !Array.isArray(meta)) {
    ["params", "params_used", "active_params", "thresholds", "strategy"].forEach((key) => {
      const value = meta[key];
      if (value && typeof value === "object" && !Array.isArray(value)) {
        candidates.push(value);
      }
    });
  }

  const hasKnown = (obj) => {
    const keys = Object.keys(obj || {}).map((k) => normalizeKey(k));
    return ["home_win_rate_threshold", "odds_min", "odds_max", "prob_threshold", "min_ev", "min_ev_per_100"].some((k) => keys.includes(k));
  };
  const matched = candidates.find(hasKnown);
  if (matched) return matched;

  return Object.fromEntries(
    Object.entries(payload ?? {}).filter(([k]) => {
      const nk = normalizeKey(k);
      return [
        "home_win_rate_threshold", "min_home_win_rate", "home_win_rate_min",
        "odds_min", "min_odds_1", "min_odds",
        "odds_max", "max_odds_1", "max_odds",
        "prob_threshold", "min_prob_used", "min_prob", "min_prob_iso",
        "min_ev", "min_ev_per_100", "min_ev_eur_per_100",
      ].includes(nk);
    }),
  );
};

export const loadStrategyParamsVersioned = (filePath) => {
  if (!fs.existsSync(filePath)) {
    return { version: 1, params: { ...DEFAULT_PARAMS }, source: "defaults" };
  }
  const payload = JSON.parse(fs.readFileSync(filePath, "utf-8"));
  const version = Number(payload?.version ?? 1);
  let raw = extractParamDict(payload);
  const params = { ...DEFAULT_PARAMS };
  Object.entries(raw).forEach(([k, v]) => {
    const n = coerceNumber(v);
    if (n !== null) {
      let nk = normalizeKey(k);
      if (["min_ev_per_100", "min_ev_eur_per_100"].includes(nk)) nk = "min_ev";
      else if (["min_home_win_rate", "home_win_rate_min"].includes(nk)) nk = "home_win_rate_threshold";
      else if (["min_odds_1", "min_odds"].includes(nk)) nk = "odds_min";
      else if (["max_odds_1", "max_odds"].includes(nk)) nk = "odds_max";
      else if (["min_prob", "min_prob_iso", "min_prob_used"].includes(nk)) nk = "prob_threshold";
      params[nk] = n;
    }
  });
  return { version, params, source: filePath };
};

export const filterLocalMatchedGamesParams = (rows, paramsObj) => {
  const params = paramsObj?.params ?? DEFAULT_PARAMS;
  return rows
    .map((row) => {
      const prob = coerceNumber(row.prob_used);
      const odds = coerceNumber(row.odds_1 ?? row.closing_home_odds);
      const ev = coerceNumber(row.ev_eur_per_100 ?? row.ev_per_100);
      const hwr = coerceNumber(row.home_win_rate);
      const gap = prob !== null && odds ? prob - 1 / odds : null;
      const blocked = [];
      if (prob === null || prob < params.prob_threshold) blocked.push("prob_threshold");
      if (odds === null || odds < params.odds_min || odds > params.odds_max) blocked.push("odds_range");
      if (ev === null || ev <= params.min_ev) blocked.push("min_ev");
      if (hwr === null || hwr < params.home_win_rate_threshold) blocked.push("home_win_rate_threshold");
      return {
        ...row,
        prob_used: prob,
        model_market_gap: gap,
        model_market_gap_flag: gap !== null ? gap > 0 : false,
        blocked_by: blocked.join("|"),
      };
    })
    .filter((row) => row.blocked_by === "");
};

export const filterLocalMatchedGamesWindow = (rows, startDate, endDate) => {
  const start = new Date(startDate);
  const end = new Date(endDate);
  return rows.filter((row) => {
    const rowDate = parseDate(row.date);
    if (!rowDate) return false;
    return rowDate >= start && rowDate <= end;
  });
};
